soc: nordic: common: dmm: Fix memory utilization

nordic-krch opened the pull request:
DMM was enforcing cache line alignment all memory regions, including those which were not cacheable. Fixing it by using memory attribute from the device tree to determine if alignment needs to be applied.

Because of that memory usage was significantly increased because even 1 byte buffers (e.g. for uart_poll_out) was consuming 32 bytes (cache line size).

Fixes #74404.

nordic-krch committed:
soc: nordic: common: dmm: Fix memory utilization

DMM was enforcing cache line alignment all memory regions, including
those which were not cacheable. Fixing it by using memory attribute
from the device tree to determine if alignment needs to be applied.

Because of that memory usage was significantly increased because
even 1 byte buffers (e.g. for uart_poll_out) was consuming 32 bytes
(cache line size).

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: boards: nrf: dmm: Adjust test to dmm changes

After changing dmm to not apply data cache line alignment for all
regions test needs to be aligned.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nika-nordic commented on soc/nordic/common/dmm.h:
```diff
@@ -35,7 +47,6 @@ extern "C" {
 #define DMM_DEV_TO_REG(node_id)					\
 	COND_CODE_1(DT_NODE_HAS_PROP(node_id, memory_regions),	\
 		((void *)DT_REG_ADDR(DT_PHANDLE(node_id, memory_regions))), (NULL))
-
```
nit: should stay I guess

nordic-krch committed:
soc: nordic: nrf54h: DMM shall be applied only to rad&app

cpuppr can only use slow peripherals and uses RAM3 as RAM so
it does not need to use DMM.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@nika-nordic @hubertmis can you take another look? I moved checking of `CONFIG_DCACHE` to `DMM_IS_REG_CACHEABLE` so that it returns 0 if dcache is disabled.

I've also added one more commit which enables DMM only for cpurad and cpuapp. Cpuppr is not using DMM.

aescolar commented:
@nordic-krch should this be merged for 3.7? (if so, could you set the milestone)
